banyan-cli/src/tags.rs

```diff
@@ -96,93 +94,94 @@ impl TreeTypes for TT {
     }
 }
 
-#[derive(Clone, Debug, Serialize, Deserialize, PartialOrd, PartialEq, Ord, Eq)]
-pub struct Tag(Arc<str>);
-
-impl Tag {
-    pub fn new(text: &str) -> Self {
-        Self(text.into())
-    }
-}
-
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
-pub struct Tags(pub BTreeSet<Tag>);
-
-impl Tags {
-    pub fn single(text: &str) -> Self {
-        Self(btreeset! { Tag(text.into()) })
-    }
+pub struct Key {
+    time: TimeData,
+    tags: TagSet,
 }
 
-#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
-pub struct Key {
+#[derive(Clone, Copy, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub struct TimeData {
     min_lamport: u64,
     min_time: u64,
     max_time: u64,
-    tags: Tags,
+}
+
+impl TimeData {
+    fn intersects(&self, that: &Self) -> bool {
+        if self.max_time < that.min_time {
+            return false;
+        }
+        if self.min_time > that.max_time {
+            return false;
+        }
+        true
+    }
+
+    fn contains(&self, that: &Self) -> bool {
+        if that.min_time < self.min_time {
+            return false;
+        }
+        if that.max_time > self.max_time {
+            return false;
+        }
+        true
+    }
+
+    fn combine(&mut self, b: &Self) {
```

Comment:
this needs a better name or some docs: it implements “union plus hole elimination” but I first expected intersection